Expose data for creation of config page programatically

joefitter opened the pull request:
- extend BaseController.prototype.locals return object with modelData from current step config
- if modelData.tableSections is present, each section is processed and a fields array of objects is created exposing value, name and step
- Custom value displaying is achieved by including a modifiers hash in the modelData object

JoeChapman commented:
@joefitter Please can you ES6 this up?

JoeChapman commented on lib/base-controller.js:
```diff
@@ -23,6 +23,41 @@ function getErrorLength() {
+ */
+function getFieldObjectsFromNames(fields, values, modifiers, steps) {
+  return fields.filter(function filter(item) {
+    return (values[item] !== undefined && values[item] !== '') || modifiers[item] !== undefined;
```
can `values[item]` be `null`?

joefitter replied:
not in any of the examples I tried, ideally it should always be `undefined` but it is set to an empty string in the /how step with the custom radio - this will probably be fixed if my PR to template mixins gets merged - https://github.com/UKHomeOffice/passports-template-mixins/pull/57

JoeChapman commented on lib/base-controller.js:
```diff
@@ -23,6 +23,41 @@ function getErrorLength() {
+  }).map(function map(item) {
+    var value = values[item];
+    var step = getStepFromFieldName(item, steps);
+    if (modifiers && modifiers[item]) {
```
Presumably `modifiers` is already known to be defined because line 46 potentially executes that test?

joefitter replied:
if `(values[item] !== undefined && values[item] !== '') === true` the `||` case is ignored. There is a test which fails if this is the case

JoeChapman replied:
My point, I suppose, is if `(values[item] !== undefined && values[item] !== '') === false` and `modifiers` === undefined`then`modifiers[item] !== undefined`will throw an exception. So perhaps you should move the test  for`modifiers` further up.

joefitter replied:
Gotcha - yeah I added the check as an afterthought to fix a failing test, I'll move it

JoeChapman commented:
@easternbloc do you have any concerns around this? Basically, it's trying to dry-up each confirmation page and remove the hard-code.

JoeChapman commented:
@easternbloc #21 is a github issue, nothing to do with our project.

easternbloc commented:
it was in the title of the PR

JoeChapman commented:
fair

easternbloc commented:
@JoeChapman @joefitter can you provide more context as to the goal of this PR?

joefitter commented:
@easternbloc there are 2 main parts to the change, firstly to expose the content of `modelData` (may be renamed `meta`) from the step config to locals, this allows variables to be defined in steps config and used in templates, goal being to remove all the copied and pasted template code and abstract to partials - example (in progress) from GRO - https://github.com/UKHomeOffice/gro/commit/c97f0df57b3eb3970745a691d44f7ff14ae37ec4

Second part of the change is to expose the data needed to create a confirm page with a partial rather than hard-coded copied and pasted code as can be seen in this commit - https://github.com/UKHomeOffice/gro/pull/15/commits/63416d24547fac454e7f05aa0c091fe9ccc866be

These are still in progress following comments from @JoeChapman so config names/locations may change resulting in changes to this PR.

easternbloc commented:
@joefitter that helps thanks 👌

joefitter commented:
Thinking about this a little more, I think the only change that should be made to BaseController is to expose `modelData` or `meta` from the steps config in the locals method.  It may be better to have a ConfirmController or SummaryController which exposes the formatted data for displaying as a table. I still think it should be part of this repo as it will be useful for every form that has a confirm step. @JoeChapman still thinking about where best to put the config for the confirm step, perhaps we could define it in a separate config file that is passed to the controller on initialisation?

JoeChapman commented on lib/confirm-controller.js:
```diff
@@ -0,0 +1,71 @@
+  Controller.apply(this, arguments);
+  if (_.isUndefined(this.options)) {
+    this.options = options || {};
+  }
```
A simpler, adequate test would be `this.options = options || {}`,

JoeChapman commented on lib/confirm-controller.js:
```diff
@@ -0,0 +1,71 @@
+  var locals = Controller.prototype.locals.apply(this, arguments);
+  var config = _.cloneDeep((this.options || {}).config) || {};
+  var tableSections = config.tableSections || [];
+  var steps = (this.options || {}).steps;
```
`this.options` has already been tested

JoeChapman commented on lib/confirm-controller.js:
```diff
@@ -0,0 +1,71 @@
+
+ConfirmController.prototype.locals = function controllerLocals(req, res) {
+  var locals = Controller.prototype.locals.apply(this, arguments);
+  var config = _.cloneDeep((this.options || {}).config) || {};
```
`this.options` has already tested

JoeChapman commented:
@easternbloc when you have a second, please can you glance at this pr?

easternbloc commented on lib/base-controller.js:
```diff
@@ -59,11 +59,13 @@ BaseController.prototype.getErrorStep = function getErrorStep(err, req) {
     nextPage: this.getNextStep(req, res),
     errorLength: getErrorLength.apply(this, arguments)
-  });
+  }, meta || {});
```
isn't this already covered in line 62?

joefitter replied:
if `this.options.meta` is `undefined` then without this check we are extending `locals` with `undefined`;

It would probably be better placed on line 62 though - `var meta = (this.options || {}).meta || {};`

easternbloc replied:
sorry yeah that's what I was actually getting at. I just used a terrible collection of words.

Perhaps `this.options` should should be tested in the constructor so that we can make the line for meta a bit cleaner. it'd clean up line 28 also?

easternbloc commented on test/lib/confirm-controller.js:
```diff
@@ -0,0 +1,108 @@
+
+      it('should contain objects', function () {
+        fields.forEach(function (field) {
+          field.should.be.an.instanceOf(Object);
```
i'm not sure about this in tests. I just prefer to do it long winded. What's testing the tests. Meh

JoeChapman commented on lib/base-controller.js:
```diff
@@ -4,8 +4,9 @@ var Controller = require('hmpo-form-wizard').Controller;
 
-var BaseController = function BaseController() {
+var BaseController = function BaseController(options) {
   this.confirmStep = '/confirm';
```
Can we `this.confirmStep = this.options.confirmStep || '/confirm';` ?

JoeChapman commented on lib/base-controller.js:
```diff
@@ -4,8 +4,9 @@ var Controller = require('hmpo-form-wizard').Controller;
+var BaseController = function BaseController(options) {
   this.confirmStep = '/confirm';
+  this.options = options || {};
   Controller.apply(this, arguments);
```
Can we 

```
this.options = options || {};
this.confirmStep = this.options.confirmStep || '/confirm';
```

JoeChapman commented on lib/confirm-controller.js:
```diff
@@ -0,0 +1,69 @@
+      res.locals.values,
+      config.modifiers || {},
+      steps
+    );
```
Can these parameters be key:values in an object?

JoeChapman commented on lib/confirm-controller.js:
```diff
@@ -0,0 +1,69 @@
+ * of field objects made up of field name,
+ * value and step
+ */
+function getFieldObjectsFromNames(fields, values, modifiers, steps) {
```
Can these arguments be key:values in an `params` object?
Using an object will alleviate some of the pain of needing to check for their presence, and simplifies the interface, i.e., which arguments are mandatory?

JoeChapman commented:
Cool, but this needs documenting.

joefitter commented:
@JoeChapman @easternbloc added docs etc, how we feeling about this PR?

JoeChapman commented on README.md:
```diff
@@ -59,6 +59,20 @@ Or override in step options
+#### Exposes meta to templates
+
+Add a meta object to step config to expose the contents of the object to the template via locals. Useful for generating templates programmatically
+
```
Add a `meta` object to a step to expose configurable key/value pairs in the template.

JoeChapman commented on README.md:
````diff
@@ -59,6 +59,20 @@ Or override in step options
+    foo: 'bar'
+  }
+}
+```
````
Could you add an example of the template, snippet will do. Please.:)

JoeChapman commented:
Looking good, please have a look at my comments. @easternbloc how do you feel about this? I think it could be a useful feature.

joefitter commented:
added to docs

JoeChapman commented:
LGTM - @easternbloc?

easternbloc commented:
@JoeChapman looking

easternbloc commented:
Does meta support translations?

I worry a little bit about moving meta to locals without checking. Can we keep the meta namespace to avoid collisions?

daniel-ac-martin commented on lib/base-controller.js:
```diff
@@ -4,8 +4,9 @@ var Controller = require('hmpo-form-wizard').Controller;
-var BaseController = function BaseController() {
-  this.confirmStep = '/confirm';
+var BaseController = function BaseController(options) {
+  this.options = options || {};
```
Pretty sure this is redundant. It will be clobbered by one of the parent constructors.
I'd suggest just referencing the options argument directly in the next line.

daniel-ac-martin replied:
See: https://github.com/UKHomeOffice/passports-form-controller/blob/master/lib/form.js#L21

JoeChapman replied:
@daniel-ac-martin do you mean the "options" argument? (s/object/options)

easternbloc replied:
He does.

easternbloc replied:
Should be `this.confirmStep = options.confirmStep || '/confirm'`

daniel-ac-martin replied:
@JoeChapman: Edited to make you look silly.

PS: Yes, thanks for pointing that out!

JoeChapman commented:
@easternbloc, it will support translations. Whether the value is translated or not is driven by the template:
Assuming:
```
meta: {
   name: 'values.one'
}
```

Then the following will be translated

`{{#t}}{{name}}{{/t}}`

easternbloc commented:
I kinda also feel like the confirm controller should be an separate PR 😈

joefitter commented:
Up to 43 comments so far - must be a record!  I can split the confirm controller to another PR if you guys prefer

joefitter commented:
![](https://sandrathorsblog.files.wordpress.com/2014/12/tumblr_llhgs0iego1qdmmwdo1_500.gif)

joefitter committed:
Closes #21 - expose data for creation of config page programatically

* Added new ConfirmController for confirmation pages, this expects a config object to be provided in step config.
* Config object is made up of tableSections array containing headers and included fields
* Custom value displaying is achieved by including a modifiers hash with keys relating to field in config object.
* Added docs to README
* Example config:

```
config: {
  tableSections: [{
    name: 'First Section',
    fields: [
      'field-1',
      'field-2'
    ]
  }, {
    name: 'Second Section',
    fields: [
      'field-3',
      'field-4
    ]
  }],
  modifiers: {
    'field-1': function(values) {
      var fields = [
        'address-text-one',
        'address-text-two',
        'address-text-three',
        'address-text-four',
        'address-text-five'
      ];
      return _.chain(values)
        .pick(fields)
        .values()
        .value()
        .join('<br />');
    }
  }
}
```

joefitter commented:
@JoeChapman @easternbloc I have split the PR in 2 - the other PR for exposing meta is https://github.com/UKHomeOffice/hof-controllers/pull/30

JoeChapman commented:
LGTM - @easternbloc can we merge this?

easternbloc commented:
@joefitter ![](https://media.giphy.com/media/NR2fifyUoGtyw/giphy.gif)
